File: heap-web/src/main/java/com/github/ol_loginov/heaplibweb/services/loaders/TypeIdLookup.java
package com.github.ol_loginov.heaplibweb.services.loaders;

import com.github.ol_loginov.heaplibweb.repository.heap.HeapRepositories;
import com.github.ol_loginov.heaplibweb.repository.heap.TypeEntity;
import lombok.RequiredArgsConstructor;

import java.util.HashMap;

@RequiredArgsConstructor
class TypeIdLookup {
	private final int heapId;
	private final HeapRepositories heapRepositories;
	private final HashMap<String, Integer> cache = new HashMap<>();

	public int lookupTypeId(String name) {
		return cache.computeIfAbsent(name, this::registerType);
	}

	private int registerType(String name) {
		var typeEntity = new TypeEntity(heapId, name);
		heapRepositories.persist(typeEntity);
		return typeEntity.getId();
	}
}

File: heap-web/src/main/java/com/github/ol_loginov/heaplibweb/services/loaders/TypeIdLookup.kt
package com.github.ol_loginov.heaplibweb.services.loaders

import com.github.ol_loginov.heaplibweb.repository.heap.HeapScope
import com.github.ol_loginov.heaplibweb.repository.heap.TypeEntity

class TypeIdLookup(
    private val heapScope: HeapScope
) {
    private val cache = HashMap<String, Int>()

    fun lookupTypeId(name: String): Int {
        return cache.computeIfAbsent(name) { n -> registerType(n) }
    }

    private fun registerType(name: String): Int {
        return heapScope.types
            .persist(TypeEntity(name))
            .id
    }
}